libs/vkd3d/breadcrumbs.c

```diff
@@ -356,9 +375,97 @@ static void vkd3d_breadcrumb_tracer_report_command_list_amd(struct vkd3d_breadcr
     ERR("Done analyzing command list.\n");
 }
 
+static void vkd3d_breadcrumb_tracer_report_command_list_nv(struct vkd3d_breadcrumb_tracer *tracer,
+        unsigned int context_index, uint32_t begin_marker, uint32_t end_marker)
+{
+    const struct vkd3d_breadcrumb_command_list_trace_context *context;
+
+    context = &tracer->trace_contexts[context_index];
+
+    /* Unused, cannot be the cause. */
+    if (context->counter == 0)
+      return;
+
+    /* Never executed, cannot be the cause. */
+    if (begin_marker == 0 && end_marker == 0)
+        return;
+
+    /* Successfully retired, cannot be the cause. */
+    if (begin_marker == UINT32_MAX && end_marker == UINT32_MAX)
+        return;
+
+    /* Edge case if we re-submitted a command list,
+     * but it ends up crashing before we hit any BOTTOM_OF_PIPE
+     * marker. Normalize the inputs such that end_marker <= begin_marker. */
+    if (begin_marker > 0 && end_marker == UINT32_MAX)
+        end_marker = 0;
+
+    ERR("Found pending command list context %u in executable state, TOP_OF_PIPE marker %u, BOTTOM_OF_PIPE marker %u.\n",
+            context_index, begin_marker, end_marker);
+    /* We set checkpoints for cmdlist begin, AFTER each command and cmdlist end.*/
+    vkd3d_breadcrumb_tracer_report_command_list(context, begin_marker, end_marker);
+    ERR("Done analyzing command list.\n");
+}
+
+static void vkd3d_breadcrumb_tracer_report_queue_nv(struct vkd3d_breadcrumb_tracer *tracer,
+                                                    struct d3d12_device *device,
+                                                    VkQueue vk_queue)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    struct vkd3d_nv_checkpoint *checkpoint;
+    VkCheckpointDataNV checkpoints[2];
+    uint32_t begin_marker, end_marker;
+    uint32_t checkpoint_count = 0;
+    uint32_t context_index;
+    uint32_t i;
+
+    VK_CALL(vkGetQueueCheckpointDataNV(vk_queue, &checkpoint_count, NULL));
+    if (checkpoint_count == 0)
+        return;
+
+    if (checkpoint_count != ARRAY_SIZE(checkpoints))
+        WARN("Driver returned more or less than 2 checkpoints.\n");
+
+    checkpoint_count = ARRAY_SIZE(checkpoints);
+    VK_CALL(vkGetQueueCheckpointDataNV(vk_queue, &checkpoint_count, checkpoints));
+
+    context_index = UINT32_MAX;
+    begin_marker = 0;
+    end_marker = 0;
+
+    for (i = 0; i < checkpoint_count; i++)
+    {
+        checkpoint = (struct vkd3d_nv_checkpoint*) checkpoints[i].pCheckpointMarker;
+        if (context_index != checkpoint->context_index && context_index != UINT32_MAX)
+        {
+            WARN("Markers have different contexts\n");
+            context_index = UINT32_MAX;
+            break;
+        }
+        context_index = checkpoint->context_index;
+        if (checkpoints[i].stage == VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT)
+            begin_marker = checkpoint->counter;
```

Comment:
Don't we want to look at first checkpoint for TOP_OF_PIPE, and last checkpoint for BOTTOM_OF_PIPE?